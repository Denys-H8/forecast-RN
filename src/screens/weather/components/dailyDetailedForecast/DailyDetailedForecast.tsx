import React, { useContext } from 'react';
import { View } from 'react-native';

import HourWeather from '../atomic/hourWeather';
import { styles } from './dailyDetailedForecast.styles';
import { timestampToHour, timestampToDate, getWeatherIconName } from 'src/helpers';
import { useSelector } from 'react-redux';
import { weatherSelector, weatherStatsSelector } from 'src/store/weather/selectors';

type Props = {
  date: string;
};

const DailyForecast: React.FC<Props> = ({ date }) => {
  const weatherStats = useSelector(weatherStatsSelector);
  const weather = useSelector(weatherSelector);

  const hourlyWeather =
    date === 'today'
      ? weatherStats.hourly.slice(0, 5)
      : weather.list
          .filter((el: { dt: number }) => {
            return timestampToDate(el.dt) === date;
          })
          .slice(0, 5);

  const getTemp = (el: { dt: number; temp: number; main: { temp: number }; weather: [{ description: string }] }) => {
    return date === 'today' ? el.temp.toFixed(1) : el.main.temp.toFixed(1);
  };

  return (
    <View style={styles.container}>
      {hourlyWeather.map(
        (el: { dt: number; temp: number; main: { temp: number }; weather: [{ description: string }] }, i: number) => {
          return (
            <HourWeather
              key={el.dt}
              hour={timestampToHour(el.dt)}
              name={getWeatherIconName(el.weather[0].description)}
              temperature={getTemp(el)}
            />
          );
        },
      )}
    </View>
  );
};

export default DailyForecast;
