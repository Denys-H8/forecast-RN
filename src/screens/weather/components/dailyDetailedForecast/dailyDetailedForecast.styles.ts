import { StyleSheet } from 'react-native';

import constants from 'src/constants';

export const styles = StyleSheet.create({
  container: {
    height: 160,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 32,
    borderTopWidth: 2,
    borderBottomWidth: 2,
    borderColor: constants.SUB_COLOR,
  },
});
