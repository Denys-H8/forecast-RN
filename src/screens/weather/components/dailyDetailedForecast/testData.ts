export const weather = {
  list: [
    {
      dt: 1619114400,
      main: {
        temp: 11.84,
        feels_like: 10.7,
        temp_min: 11.84,
        temp_max: 12,
        pressure: 1019,
        sea_level: 1019,
        grnd_level: 1002,
        humidity: 62,
        temp_kf: -0.16,
      },
      weather: [
        {
          id: 800,
          main: 'Clear',
          description: 'clear sky',
          icon: '01n',
        },
      ],
      clouds: {
        all: 0,
      },
      wind: {
        speed: 2.56,
        deg: 195,
        gust: 6.42,
      },
      visibility: 10000,
      pop: 0,
      sys: {
        pod: 'n',
      },
      dt_txt: '2021-04-22 18:00:00',
    },
    {
      dt: 1619125200,
      main: {
        temp: 11.57,
        feels_like: 10.43,
        temp_min: 11.02,
        temp_max: 11.57,
        pressure: 1019,
        sea_level: 1019,
        grnd_level: 1001,
        humidity: 63,
        temp_kf: 0.55,
      },
      weather: [
        {
          id: 802,
          main: 'Clouds',
          description: 'scattered clouds',
          icon: '03n',
        },
      ],
      clouds: {
        all: 28,
      },
      wind: {
        speed: 2.36,
        deg: 185,
        gust: 6.61,
      },
      visibility: 10000,
      pop: 0,
      sys: {
        pod: 'n',
      },
      dt_txt: '2021-04-22 21:00:00',
    },
  ],
  city: {
    id: 706483,
    name: 'Kharkiv',
    coord: {
      lat: 50,
      lon: 36.25,
    },
    country: 'UA',
    population: 1430885,
    timezone: 10800,
    sunrise: 1619058476,
    sunset: 1619109522,
  },
};

export { weatherStatsValue } from 'src/screens/forecast/components/dailyWeather/testValue';
