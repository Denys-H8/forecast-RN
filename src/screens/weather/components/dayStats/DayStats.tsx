import React from 'react';
import { View } from 'react-native';

import Humidity from '../atomic/stat';
import { styles } from './dayStats.styles';
import { getDayIndex, timestampToHour } from 'src/helpers';
import { useSelector } from 'react-redux';
import { weatherStatsSelector } from 'src/store/weather/selectors';

type Props = {
  date: string;
};

const DayStats: React.FC<Props> = ({ date }) => {
  const weatherStats = useSelector(weatherStatsSelector);

  const drops = date === 'today' ? weatherStats.current.humidity : weatherStats.daily[getDayIndex(date)].humidity;

  const wind =
    date === 'today' ? weatherStats.current.wind_speed : weatherStats.daily[getDayIndex(date)].wind_speed.toFixed(0);

  const sunrise =
    date === 'today'
      ? timestampToHour(weatherStats.current.sunrise)
      : timestampToHour(weatherStats.daily[getDayIndex(date)].sunrise);

  const sunset =
    date === 'today'
      ? timestampToHour(weatherStats.current.sunset)
      : timestampToHour(weatherStats.daily[getDayIndex(date)].sunset);

  return (
    <View style={styles.container}>
      <Humidity category="drops" text={`${drops}%`} />
      <Humidity category="wind" text={`${wind}m/c`} />
      <Humidity category="sunrise" text={sunrise} />
      <Humidity category="sunset" text={sunset} />
    </View>
  );
};

export default DayStats;
