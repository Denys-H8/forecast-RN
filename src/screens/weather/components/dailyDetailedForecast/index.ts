import DailyDetailedForecast from './DailyDetailedForecast';

export default DailyDetailedForecast;
