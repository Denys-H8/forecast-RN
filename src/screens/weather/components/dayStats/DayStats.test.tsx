import React from 'react';
import { render } from '@testing-library/react-native';
import DayStats from './DayStats';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { weatherStatsValue } from '../dailyDetailedForecast/testData';

const mockStore = configureStore([]);

const store = mockStore({
  weatherReducer: {
    weatherStats: weatherStatsValue,
  },
});

describe('DailyDetailedForecast ', () => {
  test('should render correctly with today date', () => {
    Date.now = jest.fn(() => 1619098295 * 1000);
    const date = 'today';
    const { toJSON } = render(
      <Provider store={store}>
        <DayStats date={date} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render correctly with not today date', () => {
    const date = '2021-04-22';
    const { toJSON } = render(
      <Provider store={store}>
        <DayStats date={date} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
