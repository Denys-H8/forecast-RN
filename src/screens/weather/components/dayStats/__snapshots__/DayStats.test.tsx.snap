// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyDetailedForecast  should render correctly with not today date 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 148,
      "justifyContent": "space-between",
      "paddingHorizontal": 32,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={32}
      testID="Stat.Icon"
      width={32}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      44%
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={26}
      testID="Stat.Icon"
      width={26}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      4m/c
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={28}
      testID="Stat.Icon"
      width={28}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      05:27
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={28}
      testID="Stat.Icon"
      width={28}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      19:38
    </Text>
  </View>
</View>
`;

exports[`DailyDetailedForecast  should render correctly with today date 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 148,
      "justifyContent": "space-between",
      "paddingHorizontal": 32,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={32}
      testID="Stat.Icon"
      width={32}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      32%
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={26}
      testID="Stat.Icon"
      width={26}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      6m/c
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={28}
      testID="Stat.Icon"
      width={28}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      05:27
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <IconMock
      fill="#fff"
      height={28}
      testID="Stat.Icon"
      width={28}
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginTop": 10,
          },
        ]
      }
    >
      19:38
    </Text>
  </View>
</View>
`;
